app/src/main/java/org/mozilla/focus/settings/ManualAddSearchEngineSettingsFragment.java

```diff
@@ -65,18 +112,141 @@ public boolean onOptionsItemSelected(MenuItem item) {
+        final ManualAddSearchEnginePreference pref =
+                (ManualAddSearchEnginePreference) findPreference(getString(R.string.pref_key_manual_add_search_engine));
+        pref.setProgressViewShown(isValidating);
+
+        saveMenuItem.setEnabled(!isValidating);
+    }
+
+    private static class ValidateSearchEngineAsyncTask extends AsyncTask<Void, Void, Boolean> {
+        private final WeakReference<ManualAddSearchEngineSettingsFragment> thatWeakReference;
+        private final String engineName;
+        private final String query;
+
+        private ValidateSearchEngineAsyncTask(final ManualAddSearchEngineSettingsFragment that, final String engineName,
+                final String query) {
+            this.thatWeakReference = new WeakReference<>(that);
+            this.engineName = engineName;
+            this.query = query;
+        }
+
+        @Override
+        protected Boolean doInBackground(final Void... voids) {
+            final boolean isValidSearchQuery = isValidSearchQueryURL(query);
+            TelemetryWrapper.saveCustomSearchEngineEvent(isValidSearchQuery);
+            return isValidSearchQuery;
+        }
+
+        @Override
+        protected void onPostExecute(final Boolean isValidSearchQuery) {
+            super.onPostExecute(isValidSearchQuery);
+
+            if (isCancelled()) {
+                Log.d(LOGTAG, "ValidateSearchEngineAsyncTask has been cancelled");
+                return;
+            }
+
+            final ManualAddSearchEngineSettingsFragment that = thatWeakReference.get();
+            if (that == null) {
+                Log.d(LOGTAG, "Fragment or menu item no longer exists when search query validation async task returned.");
+                return;
+            }
+
+            if (isValidSearchQuery) {
+                final SharedPreferences sharedPreferences = that.getSearchEngineSharedPreferences();
+                SearchEngineManager.addSearchEngine(sharedPreferences, that.getActivity(), engineName, query);
+                Snackbar.make(that.getView(), R.string.search_add_confirmation, Snackbar.LENGTH_SHORT).show();
+                that.getFragmentManager().popBackStack();
+            } else {
+                showServerError(that);
+            }
+
+            that.setUiIsValidatingAsync(false, that.menuItemForActiveAsyncTask);
+            that.activeAsyncTask = null;
+            that.menuItemForActiveAsyncTask = null;
+        }
+
+        private void showServerError(final ManualAddSearchEngineSettingsFragment that) {
+            final ManualAddSearchEnginePreference pref = (ManualAddSearchEnginePreference) that.findPreference(
+                    that.getString(R.string.pref_key_manual_add_search_engine));
+            pref.setSearchQueryErrorText(that.getString(R.string.error_hostLookup_title));
+        }
+    }
+
+    @WorkerThread // makes network request.
+    private static boolean isValidSearchQueryURL(final String query) {
```

Comment:
\[PMD\] INFO: The method 'isValidSearchQueryURL' has a Cyclomatic Complexity of 5\.